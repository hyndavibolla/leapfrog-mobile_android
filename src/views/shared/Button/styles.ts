import { css } from '@emotion/native';

import { COLOR, FONT_SIZE } from '../../../constants';

export const styles = {
  container: css``,
  innerContainer: css`
    padding: 8px 12px;
    background-color: ${COLOR.PRIMARY_BLUE};
    border-radius: 30px;
    flex-direction: row;
    align-items: center;
    justify-content: center;
  `,
  childContainerPadding: css`
    margin-left: 8px;
  `,
  text: css`
    color: ${COLOR.WHITE};
    font-size: ${FONT_SIZE.PETITE};
    text-align: center;
  `,
  disabled: css`
    opacity: 0.5;
  `
};
