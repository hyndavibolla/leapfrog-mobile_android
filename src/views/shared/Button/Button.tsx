import React, { memo, ReactElement, ReactNode, useMemo, ComponentProps } from 'react';
import { StyleProp, TextStyle, View, ViewStyle, TouchableOpacity } from 'react-native';
/** this is needed from "react-native-gesture-handler" for compatibility with other buggy libraries */
import { TouchableOpacity as TouchableOpacityAlt } from 'react-native-gesture-handler';

import { styles } from './styles';
import { Text } from '../Text';

type TouchableOpacityProps = ComponentProps<typeof TouchableOpacity>;

export interface Props extends TouchableOpacityProps {
  children: (string | number | ((index: number) => ReactElement))[] | (string | number | ReactElement | ReactNode);
  style?: StyleProp<ViewStyle>;
  containerColor?: string;
  innerContainerStyle?: StyleProp<ViewStyle>;
  innerContainerDisabledStyle?: StyleProp<ViewStyle>;
  textStyle?: StyleProp<TextStyle>;
  textColor?: string;
  compatibilityMode?: boolean;
}

export const Button = ({
  children,
  style,
  containerColor,
  textColor,
  textStyle,
  innerContainerStyle,
  innerContainerDisabledStyle,
  disabled,
  compatibilityMode,
  ...props
}: Props) => {
  const containerStyleList = useMemo(() => [styles.container, style], [style]);
  const innerContainerStyleList = useMemo(
    () =>
      [
        styles.innerContainer,
        innerContainerStyle,
        containerColor && { backgroundColor: containerColor },
        disabled && styles.disabled,
        disabled && innerContainerDisabledStyle
      ].filter(Boolean),
    [innerContainerStyle, containerColor, disabled, innerContainerDisabledStyle]
  );
  const textStyleList = useMemo(() => [styles.text, textStyle, textColor && { color: textColor }], [textStyle, textColor]);

  const Touchable = (compatibilityMode ? TouchableOpacityAlt : TouchableOpacity) as typeof TouchableOpacity;

  return (
    <Touchable {...props} style={containerStyleList} disabled={disabled}>
      <View style={innerContainerStyleList}>
        {((children instanceof Array ? children : [children]) as ReactNode[]).map((child, index) => (
          <View key={index} style={!!index && styles.childContainerPadding}>
            {(() => {
              const type = typeof child;
              if (type === 'function') return (child as (i: number) => ReactNode)(index);
              if (['string', 'number'].includes(type)) return <Text style={textStyleList}>{child}</Text>;
              return child;
            })()}
          </View>
        ))}
      </View>
    </Touchable>
  );
};

export default memo(Button);
