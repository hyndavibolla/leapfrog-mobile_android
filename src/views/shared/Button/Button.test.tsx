import React from 'react';
import { fireEvent, render } from '@testing-library/react-native';

import Button, { Props } from './Button';

describe('Button', () => {
  let props: Props;
  beforeEach(() => {
    props = {
      onPress: jest.fn(),
      children: 'default content',
      testID: 'btn'
    };
  });

  it('should render with a single string child', () => {
    const { toJSON } = render(<Button {...props}>Regular button</Button>);
    expect(toJSON()).toMatchSnapshot();
  });

  it('should render on compatibility mode', () => {
    const { toJSON } = render(
      <Button {...props} compatibilityMode={true}>
        Regular button
      </Button>
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('should render with a single node child', () => {
    const { toJSON } = render(
      <Button {...props}>
        <div>node child</div>
      </Button>
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('should render with many children', () => {
    const { toJSON } = render(<Button {...props}>{['string child', () => <div>node child</div>]}</Button>);
    expect(toJSON()).toMatchSnapshot();
  });

  it('should render with customizations', () => {
    const { toJSON } = render(
      <Button
        {...props}
        style={{ margin: 1 }}
        containerColor="red"
        innerContainerStyle={{ margin: 2 }}
        textStyle={{ margin: 3 }}
        textColor="blue"
        disabled={true}
        innerContainerDisabledStyle={{ opacity: 1 }}
      >
        Regular button
      </Button>
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('should be pressable', () => {
    const { getByTestId } = render(<Button {...props}>Regular button</Button>);
    fireEvent.press(getByTestId('btn'));
    expect(props.onPress).toBeCalled();
  });
});
