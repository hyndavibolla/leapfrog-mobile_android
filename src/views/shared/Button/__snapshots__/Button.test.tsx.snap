// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render on compatibility mode 1`] = `
<RNGestureHandlerButton
  collapsable={false}
  onGestureEvent={[Function]}
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
  onHandlerStateChange={[Function]}
  rippleColor={0}
  testID="btn"
>
  <View
    accessible={true}
    collapsable={false}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#0066CC",
            "borderBottomLeftRadius": 30,
            "borderBottomRightRadius": 30,
            "borderTopLeftRadius": 30,
            "borderTopRightRadius": 30,
            "flexDirection": "row",
            "justifyContent": "center",
            "paddingBottom": 8,
            "paddingLeft": 12,
            "paddingRight": 12,
            "paddingTop": 8,
          },
        ]
      }
    >
      <View
        style={false}
      >
        <Text
          style={
            Array [
              Object {
                "color": "#909BA0",
                "fontFamily": "SFProDisplay-Medium",
                "fontSize": 16,
                "lineHeight": 22,
              },
              Array [
                Object {
                  "color": "#FFFFFF",
                  "fontSize": 14,
                  "textAlign": "center",
                },
                undefined,
                undefined,
              ],
            ]
          }
        >
          Regular button
        </Text>
      </View>
    </View>
  </View>
</RNGestureHandlerButton>
`;

exports[`Button should render with a single node child 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID="btn"
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#0066CC",
          "borderBottomLeftRadius": 30,
          "borderBottomRightRadius": 30,
          "borderTopLeftRadius": 30,
          "borderTopRightRadius": 30,
          "flexDirection": "row",
          "justifyContent": "center",
          "paddingBottom": 8,
          "paddingLeft": 12,
          "paddingRight": 12,
          "paddingTop": 8,
        },
      ]
    }
  >
    <View
      style={false}
    >
      <div>
        node child
      </div>
    </View>
  </View>
</View>
`;

exports[`Button should render with a single string child 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID="btn"
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#0066CC",
          "borderBottomLeftRadius": 30,
          "borderBottomRightRadius": 30,
          "borderTopLeftRadius": 30,
          "borderTopRightRadius": 30,
          "flexDirection": "row",
          "justifyContent": "center",
          "paddingBottom": 8,
          "paddingLeft": 12,
          "paddingRight": 12,
          "paddingTop": 8,
        },
      ]
    }
  >
    <View
      style={false}
    >
      <Text
        style={
          Array [
            Object {
              "color": "#909BA0",
              "fontFamily": "SFProDisplay-Medium",
              "fontSize": 16,
              "lineHeight": 22,
            },
            Array [
              Object {
                "color": "#FFFFFF",
                "fontSize": 14,
                "textAlign": "center",
              },
              undefined,
              undefined,
            ],
          ]
        }
      >
        Regular button
      </Text>
    </View>
  </View>
</View>
`;

exports[`Button should render with customizations 1`] = `
<View
  accessibilityState={
    Object {
      "disabled": true,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "margin": 1,
      "opacity": 1,
    }
  }
  testID="btn"
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#0066CC",
          "borderBottomLeftRadius": 30,
          "borderBottomRightRadius": 30,
          "borderTopLeftRadius": 30,
          "borderTopRightRadius": 30,
          "flexDirection": "row",
          "justifyContent": "center",
          "paddingBottom": 8,
          "paddingLeft": 12,
          "paddingRight": 12,
          "paddingTop": 8,
        },
        Object {
          "margin": 2,
        },
        Object {
          "backgroundColor": "red",
        },
        Object {
          "opacity": 0.5,
        },
        Object {
          "opacity": 1,
        },
      ]
    }
  >
    <View
      style={false}
    >
      <Text
        style={
          Array [
            Object {
              "color": "#909BA0",
              "fontFamily": "SFProDisplay-Medium",
              "fontSize": 16,
              "lineHeight": 22,
            },
            Array [
              Object {
                "color": "#FFFFFF",
                "fontSize": 14,
                "textAlign": "center",
              },
              Object {
                "margin": 3,
              },
              Object {
                "color": "blue",
              },
            ],
          ]
        }
      >
        Regular button
      </Text>
    </View>
  </View>
</View>
`;

exports[`Button should render with many children 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID="btn"
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#0066CC",
          "borderBottomLeftRadius": 30,
          "borderBottomRightRadius": 30,
          "borderTopLeftRadius": 30,
          "borderTopRightRadius": 30,
          "flexDirection": "row",
          "justifyContent": "center",
          "paddingBottom": 8,
          "paddingLeft": 12,
          "paddingRight": 12,
          "paddingTop": 8,
        },
      ]
    }
  >
    <View
      style={false}
    >
      <Text
        style={
          Array [
            Object {
              "color": "#909BA0",
              "fontFamily": "SFProDisplay-Medium",
              "fontSize": 16,
              "lineHeight": 22,
            },
            Array [
              Object {
                "color": "#FFFFFF",
                "fontSize": 14,
                "textAlign": "center",
              },
              undefined,
              undefined,
            ],
          ]
        }
      >
        string child
      </Text>
    </View>
    <View
      style={
        Object {
          "marginLeft": 8,
        }
      }
    >
      <div>
        node child
      </div>
    </View>
  </View>
</View>
`;
